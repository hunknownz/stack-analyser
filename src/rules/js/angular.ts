import { register } from '../../rules.js';

register({
  tech: 'angular',
  files: ['angular.json'],
  dependencies: [
    { type: 'npm', name: 'angular' },
    { type: 'npm', name: /^@angular\//, example: '@angular/forms' },
  ],
});
