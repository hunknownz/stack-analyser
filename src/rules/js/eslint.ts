import { register } from '../../register.js';

register({
  tech: 'eslint',
  name: 'Eslint',
  type: 'tool',
  files: [
    '.eslintrc',
    '.eslintrc.cjs',
    '.eslintrc.json',
    '.eslintrc.js',
    'eslint.config.js',
    'eslint.config.cjs',
    'eslint.config.mjs',
  ],
  dependencies: [{ type: 'npm', name: 'eslint' }],
});
