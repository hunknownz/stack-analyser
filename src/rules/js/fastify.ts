import { register } from '../../register.js';

register({
  tech: 'fastify',
  name: 'Fastify',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'fastify' }],
});
