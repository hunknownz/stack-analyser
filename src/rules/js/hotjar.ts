import { register } from '../../register.js';

register({
  tech: 'hotjar',
  name: 'Hotjar',
  type: 'tool',
  dependencies: [{ type: 'npm', name: '@hotjar/browser' }],
});
