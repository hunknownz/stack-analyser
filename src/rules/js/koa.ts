import { register } from '../../register.js';

register({
  tech: 'koa',
  name: 'Koa',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'koa' }],
});
