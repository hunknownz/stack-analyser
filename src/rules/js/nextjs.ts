import { register } from '../../register.js';

register({
  tech: 'nextjs',
  name: 'Next.js',
  type: 'framework',
  files: ['next.config.js'],
  dependencies: [{ type: 'npm', name: 'next' }],
});
