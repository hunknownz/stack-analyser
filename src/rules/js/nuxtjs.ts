import { register } from '../../register.js';

register({
  tech: 'nuxtjs',
  name: 'Nuxt.js',
  type: 'framework',
  files: ['nuxt.config.js'],
  dependencies: [{ type: 'npm', name: 'nuxt' }],
});
