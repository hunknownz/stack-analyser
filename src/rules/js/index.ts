import './angular.js';
import './bootstrap.js';
import './esbuild.js';
import './eslint.js';
import './express.js';
import './fastify.js';
import './hotjar.js';
import './koa.js';
import './prettier.js';
import './prisma.js';
import './react.js';
import './reactEmail.js';
import './rollup.js';
import './sequelize.js';
import './socketio.js';
import './storybook.js';
import './strapi.js';
import './stylelint.js';
import './tailwind.js';
import './typescript.js';
import './vite.js';
import './vue.js';
import './webpack.js';
