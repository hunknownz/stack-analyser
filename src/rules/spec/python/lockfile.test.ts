import { describe, expect, it } from 'vitest';

import { analyser } from '../../../analyser/index.js';
import { rawList } from '../../../loader.js';
import { flatten } from '../../../payload/helpers.js';
import { FakeProvider } from '../../../provider/fake.js';

import '../../../autoload.js';

describe('python lockfile: requirements.txt', () => {
  it('should match everything', async () => {
    const deps = [];
    for (const item of rawList) {
      if (item.type !== 'dependency' || item.ref.type !== 'python') {
        continue;
      }

      const example = 'example' in item.ref ? item.ref.example : item.ref.name;
      deps.push(`${example}==1.0.0`);
    }

    const lockfile: string = deps.join('\n');

    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['requirements.txt'],
        },
        files: {
          '/requirements.txt': lockfile,
        },
      }),
    });

    const merged = flatten(res, { merge: true });

    expect([...merged.techs].sort()).toMatchSnapshot();
    expect([...merged.dependencies].sort()).toMatchSnapshot();
  });

  it('should match weird syntax', async () => {
    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['requirements.txt'],
        },
        files: {
          '/requirements.txt': [
            // No version
            'tensorflow',
            // gte
            'algoliasearch>=1.9.1',
            // gte-lt
            'datadog>=1.4.3,<7.0.0',
          ].join('\n'),
        },
      }),
    });

    const merged = flatten(res, { merge: true });

    expect([...merged.techs].sort()).toMatchSnapshot();
    expect([...merged.dependencies].sort()).toMatchSnapshot();
  });
});

describe('python lockfile: pyproject.toml', () => {
  it('should match everything', async () => {
    const deps = [];
    for (const item of rawList) {
      if (item.type !== 'dependency' || item.ref.type !== 'python') {
        continue;
      }

      const example = 'example' in item.ref ? item.ref.example : item.ref.name;
      deps.push(`"${example}==1.0.0"`);
    }

    const lockfile = `[project]
name = "test"
description = "This is a test"

dependencies = [
${deps.join(',\r\n')}
]

[project.optional-dependencies]
memory = [
    "sentence-transformers>=4.0.2",
]

[tool.ruff]
line-length = 130
`;

    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['pyproject.toml'],
        },
        files: {
          '/pyproject.toml': lockfile,
        },
      }),
    });

    const merged = flatten(res, { merge: true });

    expect([...merged.techs].sort()).toMatchSnapshot();
    expect([...merged.dependencies].sort()).toMatchSnapshot();
  });

  it('should match license', async () => {
    const lockfile = `[project]
name = "test"
license = "ISC"
`;

    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['pyproject.toml'],
        },
        files: {
          '/pyproject.toml': lockfile,
        },
      }),
    });

    const merged = flatten(res, { merge: true });

    expect([...merged.licenses.values()]).toStrictEqual(['ISC']);
  });
});
