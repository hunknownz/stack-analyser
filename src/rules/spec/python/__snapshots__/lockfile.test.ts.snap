// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`python lockfile: pyproject.toml > should match everything 1`] = `
[
  "algolia",
  "anthropic",
  "apideck",
  "browerbase",
  "celery",
  "chargebee",
  "chromadb",
  "clerk",
  "clickhouse",
  "cohereai",
  "convexdb",
  "datadog",
  "deepseek",
  "elasticsearch",
  "firecrawl",
  "frontegg",
  "fusionauth",
  "gcp",
  "gcp.gcs",
  "geminiai",
  "groq",
  "hetzner",
  "hostinger",
  "huggingface",
  "inngest",
  "kinde",
  "lago",
  "libsql",
  "logtoio",
  "mergedev",
  "milvusdb",
  "mistralai",
  "mixpanel",
  "mongodb",
  "mysql",
  "neondb",
  "novu",
  "ollama",
  "openai",
  "paddle",
  "pinecone",
  "posthog",
  "qdrant",
  "resend",
  "stytch",
  "supabase",
  "supertokens",
  "swiftype",
  "withorb",
  "workos",
  "xai",
  "zuora",
]
`;

exports[`python lockfile: pyproject.toml > should match everything 2`] = `
[
  [
    "python",
    "MergePythonClient",
    "1.0.0",
  ],
  [
    "python",
    "MergePythonSDK",
    "1.0.0",
  ],
  [
    "python",
    "PyMySQL",
    "1.0.0",
  ],
  [
    "python",
    "SQLAlchemy",
    "1.0.0",
  ],
  [
    "python",
    "algoliasearch",
    "1.0.0",
  ],
  [
    "python",
    "algoliasearch-django",
    "1.0.0",
  ],
  [
    "python",
    "anthropic",
    "1.0.0",
  ],
  [
    "python",
    "apideck-unify",
    "1.0.0",
  ],
  [
    "python",
    "browser-use",
    "1.0.0",
  ],
  [
    "python",
    "browserbase",
    "1.0.0",
  ],
  [
    "python",
    "celery",
    "1.0.0",
  ],
  [
    "python",
    "chargebee",
    "1.0.0",
  ],
  [
    "python",
    "chromadb",
    "1.0.0",
  ],
  [
    "python",
    "clerk-backend-api",
    "1.0.0",
  ],
  [
    "python",
    "clickhouse-connect",
    "1.0.0",
  ],
  [
    "python",
    "cohere",
    "1.0.0",
  ],
  [
    "python",
    "convex",
    "1.0.0",
  ],
  [
    "python",
    "datadog",
    "1.0.0",
  ],
  [
    "python",
    "django",
    "1.0.0",
  ],
  [
    "python",
    "django-workos",
    "1.0.0",
  ],
  [
    "python",
    "elastic-site-search",
    "1.0.0",
  ],
  [
    "python",
    "elasticsearch",
    "1.0.0",
  ],
  [
    "python",
    "firecrawl-py",
    "1.0.0",
  ],
  [
    "python",
    "frontegg",
    "1.0.0",
  ],
  [
    "python",
    "fusionauth-client",
    "1.0.0",
  ],
  [
    "python",
    "genai",
    "1.0.0",
  ],
  [
    "python",
    "google-cloud-storage",
    "1.0.0",
  ],
  [
    "python",
    "gradio",
    "1.0.0",
  ],
  [
    "python",
    "hcloud",
    "1.0.0",
  ],
  [
    "python",
    "hostinger_api",
    "1.0.0",
  ],
  [
    "python",
    "inngest",
    "1.0.0",
  ],
  [
    "python",
    "kinde-python-sdk",
    "1.0.0",
  ],
  [
    "python",
    "lago-python-client",
    "1.0.0",
  ],
  [
    "python",
    "langchain-anthropic",
    "1.0.0",
  ],
  [
    "python",
    "langchain-deepseek",
    "1.0.0",
  ],
  [
    "python",
    "langchain-google-genai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-groq",
    "1.0.0",
  ],
  [
    "python",
    "langchain-huggingface",
    "1.0.0",
  ],
  [
    "python",
    "langchain-milvus",
    "1.0.0",
  ],
  [
    "python",
    "langchain-mistralai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-ollama",
    "1.0.0",
  ],
  [
    "python",
    "langchain-openai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-xai",
    "1.0.0",
  ],
  [
    "python",
    "libsql-experimental",
    "1.0.0",
  ],
  [
    "python",
    "logto",
    "1.0.0",
  ],
  [
    "python",
    "mistralai",
    "1.0.0",
  ],
  [
    "python",
    "mixpanel",
    "1.0.0",
  ],
  [
    "python",
    "mysql-connector-python",
    "1.0.0",
  ],
  [
    "python",
    "mysqlclient",
    "1.0.0",
  ],
  [
    "python",
    "neon-api",
    "1.0.0",
  ],
  [
    "python",
    "novu",
    "1.0.0",
  ],
  [
    "python",
    "openai",
    "1.0.0",
  ],
  [
    "python",
    "orb-billing",
    "1.0.0",
  ],
  [
    "python",
    "paddle-python-sdk",
    "1.0.0",
  ],
  [
    "python",
    "pinecone",
    "1.0.0",
  ],
  [
    "python",
    "posthog",
    "1.0.0",
  ],
  [
    "python",
    "pulumi",
    "1.0.0",
  ],
  [
    "python",
    "pymilvus",
    "1.0.0",
  ],
  [
    "python",
    "pymongo",
    "1.0.0",
  ],
  [
    "python",
    "pytest-playwright",
    "1.0.0",
  ],
  [
    "python",
    "python-socketio",
    "1.0.0",
  ],
  [
    "python",
    "pytorch",
    "1.0.0",
  ],
  [
    "python",
    "qdrant-client",
    "1.0.0",
  ],
  [
    "python",
    "resend",
    "1.0.0",
  ],
  [
    "python",
    "selenium",
    "1.0.0",
  ],
  [
    "python",
    "stytch",
    "1.0.0",
  ],
  [
    "python",
    "supabase",
    "1.0.0",
  ],
  [
    "python",
    "supertokens_python",
    "1.0.0",
  ],
  [
    "python",
    "tensorflow",
    "1.0.0",
  ],
  [
    "python",
    "workos",
    "1.0.0",
  ],
  [
    "python",
    "zuora-sdk",
    "1.0.0",
  ],
]
`;

exports[`python lockfile: requirements.txt > should match everything 1`] = `
[
  "algolia",
  "anthropic",
  "apideck",
  "browerbase",
  "celery",
  "chargebee",
  "chromadb",
  "clerk",
  "clickhouse",
  "cohereai",
  "convexdb",
  "datadog",
  "deepseek",
  "elasticsearch",
  "firecrawl",
  "frontegg",
  "fusionauth",
  "gcp",
  "gcp.gcs",
  "geminiai",
  "groq",
  "hetzner",
  "hostinger",
  "huggingface",
  "inngest",
  "kinde",
  "lago",
  "libsql",
  "logtoio",
  "mergedev",
  "milvusdb",
  "mistralai",
  "mixpanel",
  "mongodb",
  "mysql",
  "neondb",
  "novu",
  "ollama",
  "openai",
  "paddle",
  "pinecone",
  "posthog",
  "python",
  "qdrant",
  "resend",
  "stytch",
  "supabase",
  "supertokens",
  "swiftype",
  "withorb",
  "workos",
  "xai",
  "zuora",
]
`;

exports[`python lockfile: requirements.txt > should match everything 2`] = `
[
  [
    "python",
    "MergePythonClient",
    "1.0.0",
  ],
  [
    "python",
    "MergePythonSDK",
    "1.0.0",
  ],
  [
    "python",
    "PyMySQL",
    "1.0.0",
  ],
  [
    "python",
    "SQLAlchemy",
    "1.0.0",
  ],
  [
    "python",
    "algoliasearch",
    "1.0.0",
  ],
  [
    "python",
    "algoliasearch-django",
    "1.0.0",
  ],
  [
    "python",
    "anthropic",
    "1.0.0",
  ],
  [
    "python",
    "apideck-unify",
    "1.0.0",
  ],
  [
    "python",
    "browser-use",
    "1.0.0",
  ],
  [
    "python",
    "browserbase",
    "1.0.0",
  ],
  [
    "python",
    "celery",
    "1.0.0",
  ],
  [
    "python",
    "chargebee",
    "1.0.0",
  ],
  [
    "python",
    "chromadb",
    "1.0.0",
  ],
  [
    "python",
    "clerk-backend-api",
    "1.0.0",
  ],
  [
    "python",
    "clickhouse-connect",
    "1.0.0",
  ],
  [
    "python",
    "cohere",
    "1.0.0",
  ],
  [
    "python",
    "convex",
    "1.0.0",
  ],
  [
    "python",
    "datadog",
    "1.0.0",
  ],
  [
    "python",
    "django",
    "1.0.0",
  ],
  [
    "python",
    "django-workos",
    "1.0.0",
  ],
  [
    "python",
    "elastic-site-search",
    "1.0.0",
  ],
  [
    "python",
    "elasticsearch",
    "1.0.0",
  ],
  [
    "python",
    "firecrawl-py",
    "1.0.0",
  ],
  [
    "python",
    "frontegg",
    "1.0.0",
  ],
  [
    "python",
    "fusionauth-client",
    "1.0.0",
  ],
  [
    "python",
    "genai",
    "1.0.0",
  ],
  [
    "python",
    "google-cloud-storage",
    "1.0.0",
  ],
  [
    "python",
    "gradio",
    "1.0.0",
  ],
  [
    "python",
    "hcloud",
    "1.0.0",
  ],
  [
    "python",
    "hostinger_api",
    "1.0.0",
  ],
  [
    "python",
    "inngest",
    "1.0.0",
  ],
  [
    "python",
    "kinde-python-sdk",
    "1.0.0",
  ],
  [
    "python",
    "lago-python-client",
    "1.0.0",
  ],
  [
    "python",
    "langchain-anthropic",
    "1.0.0",
  ],
  [
    "python",
    "langchain-deepseek",
    "1.0.0",
  ],
  [
    "python",
    "langchain-google-genai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-groq",
    "1.0.0",
  ],
  [
    "python",
    "langchain-huggingface",
    "1.0.0",
  ],
  [
    "python",
    "langchain-milvus",
    "1.0.0",
  ],
  [
    "python",
    "langchain-mistralai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-ollama",
    "1.0.0",
  ],
  [
    "python",
    "langchain-openai",
    "1.0.0",
  ],
  [
    "python",
    "langchain-xai",
    "1.0.0",
  ],
  [
    "python",
    "libsql-experimental",
    "1.0.0",
  ],
  [
    "python",
    "logto",
    "1.0.0",
  ],
  [
    "python",
    "mistralai",
    "1.0.0",
  ],
  [
    "python",
    "mixpanel",
    "1.0.0",
  ],
  [
    "python",
    "mysql-connector-python",
    "1.0.0",
  ],
  [
    "python",
    "mysqlclient",
    "1.0.0",
  ],
  [
    "python",
    "neon-api",
    "1.0.0",
  ],
  [
    "python",
    "novu",
    "1.0.0",
  ],
  [
    "python",
    "openai",
    "1.0.0",
  ],
  [
    "python",
    "orb-billing",
    "1.0.0",
  ],
  [
    "python",
    "paddle-python-sdk",
    "1.0.0",
  ],
  [
    "python",
    "pinecone",
    "1.0.0",
  ],
  [
    "python",
    "posthog",
    "1.0.0",
  ],
  [
    "python",
    "pulumi",
    "1.0.0",
  ],
  [
    "python",
    "pymilvus",
    "1.0.0",
  ],
  [
    "python",
    "pymongo",
    "1.0.0",
  ],
  [
    "python",
    "pytest-playwright",
    "1.0.0",
  ],
  [
    "python",
    "python-socketio",
    "1.0.0",
  ],
  [
    "python",
    "pytorch",
    "1.0.0",
  ],
  [
    "python",
    "qdrant-client",
    "1.0.0",
  ],
  [
    "python",
    "resend",
    "1.0.0",
  ],
  [
    "python",
    "selenium",
    "1.0.0",
  ],
  [
    "python",
    "stytch",
    "1.0.0",
  ],
  [
    "python",
    "supabase",
    "1.0.0",
  ],
  [
    "python",
    "supertokens_python",
    "1.0.0",
  ],
  [
    "python",
    "tensorflow",
    "1.0.0",
  ],
  [
    "python",
    "workos",
    "1.0.0",
  ],
  [
    "python",
    "zuora-sdk",
    "1.0.0",
  ],
]
`;

exports[`python lockfile: requirements.txt > should match weird syntax 1`] = `
[
  "algolia",
  "datadog",
  "python",
]
`;

exports[`python lockfile: requirements.txt > should match weird syntax 2`] = `
[
  [
    "python",
    "algoliasearch",
    "1.9.1",
  ],
  [
    "python",
    "datadog",
    "1.4.3",
  ],
  [
    "python",
    "tensorflow",
    "latest",
  ],
]
`;
